File: projects/tax-id/src/lib/countries/albania.ts
import { TaxIdValidationResult } from '../models';
import { normalizeTaxId } from '../normalize';

export function validateAlbanianTaxId(value: unknown): TaxIdValidationResult {
  const normalizedValue = normalizeTaxId(value);
  const base = { country: 'AL', normalizedValue } as const;

  if (!normalizedValue) {
    return { ...base, valid: false, error: 'empty' };
  }

  if (normalizedValue.length !== 10) {
    return { ...base, valid: false, error: 'invalid_length' };
  }

  if (!/^(?:\d{2}|[A-T]\d)\d{7}[A-W]$/.test(normalizedValue)) {
    return { ...base, valid: false, error: 'invalid_format' };
  }

  const year = decodeYear(normalizedValue.slice(0, 2));
  const encodedMonth = Number(normalizedValue.slice(2, 4));
  const month = decodeMonth(encodedMonth);
  const day = Number(normalizedValue.slice(4, 6));
  const serial = Number(normalizedValue.slice(6, 9));

  if (year === null || month === null || serial === 0 || !isValidDate(year, month, day)) {
    return { ...base, valid: false, error: 'invalid_format' };
  }

  return { ...base, valid: true, validationLevel: 'format' };
}

function decodeYear(value: string): number | null {
  if (/^\d{2}$/.test(value)) {
    return 1800 + Number(value);
  }

  const decade = value.charCodeAt(0) - 'A'.charCodeAt(0);
  const yearInDecade = Number(value[1]);

  if (decade >= 0 && decade <= 9) {
    return 1900 + decade * 10 + yearInDecade;
  }

  if (decade >= 10 && decade <= 19) {
    return 2000 + (decade - 10) * 10 + yearInDecade;
  }

  return null;
}

function decodeMonth(value: number): number | null {
  for (const offset of [0, 30, 50, 80]) {
    const month = value - offset;
    if (month >= 1 && month <= 12) {
      return month;
    }
  }

  return null;
}

function isValidDate(year: number, month: number, day: number): boolean {
  const date = new Date(Date.UTC(year, month - 1, day));
  return (
    date.getUTCFullYear() === year &&
    date.getUTCMonth() === month - 1 &&
    date.getUTCDate() === day
  );
}

File: packages/dotnet/NationalIdentifiers.Core/Countries/Albania.cs
using System.Text.RegularExpressions;
namespace NationalIdentifiers.Core.Countries;
internal static class Albania
{
    internal static ValidationResult Validate(object? value)
    {
        var n = TaxIdNormalizer.Normalize(value);
        if (string.IsNullOrEmpty(n)) return ValidationResult.Fail("AL", n, ValidationErrorCode.Empty);
        if (n.Length != 10) return ValidationResult.Fail("AL", n, ValidationErrorCode.InvalidLength);
        if (!Regex.IsMatch(n, @"^(?:\d{2}|[A-T]\d)\d{7}[A-W]$")) return ValidationResult.Fail("AL", n, ValidationErrorCode.InvalidFormat);
        int? yr = DecodeYear(n[..2]);
        int em = int.Parse(n[2..4]); int dy = int.Parse(n[4..6]);
        int? mo = DecodeMonth(em);
        int serial = int.Parse(n[6..9]);
        if (yr == null || mo == null || serial == 0 || !DateValidator.IsValidDate(yr.Value, mo.Value, dy))
            return ValidationResult.Fail("AL", n, ValidationErrorCode.InvalidFormat);
        return ValidationResult.Ok("AL", n, ValidationLevel.Format);
    }
    private static int? DecodeYear(string v)
    {
        if (Regex.IsMatch(v, @"^\d{2}$")) return 1800 + int.Parse(v);
        int decade = v[0] - 'A'; int y = v[1] - '0';
        if (decade >= 0 && decade <= 9) return 1900 + decade*10 + y;
        if (decade >= 10 && decade <= 19) return 2000 + (decade-10)*10 + y;
        return null;
    }
    private static int? DecodeMonth(int em)
    {
        foreach (int offset in new[]{0,30,50,80})
        { int m = em - offset; if (m >= 1 && m <= 12) return m; }
        return null;
    }
}
